llvm/test/CodeGen/NVPTX/kernel-param-align.ll

```diff
@@ -0,0 +1,34 @@
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_72 2>&1 | FileCheck %s
+; RUN: %if ptxas %{ llc < %s -march=nvptx64 -mcpu=sm_72 | %ptxas-verify %}
+
+%struct.Large = type { [16 x double] }
+
+; CHECK: .param .u64 .ptr .global .align 16 func_align_param_0,
+; CHECK: .param .u64 func_align_param_1,
```

Comment:
.ptr isn't annotated for second argument in CUDA as seen in the code. Currently for CUDA, we are only supporting .global .ptr for global memory objects in this scope of PR for CUDA. 

@Artem-B there are 2 issues with this code that I had overseen. 
1. For the second argument in both cases, the current code incorrectly is showing .global .ptr even though the pointer is not pointing to objects in global memory space since we have no check for global memory space apart from CUDA check.
2. For the same second argument in both the cases, I see .global .ptr .align 0, the alignment specified as 0 since we are assigning alignment with getParamAlign() that assumes alignment is specified. 

To resolve both the issues, I think we can stick with the initial scope of this PR to assign .global .ptr to pointers whose alignment is specified as ptr nocapture readonly align 16. For this, I think reverting this change to the initial commit - [https://github.com/llvm/llvm-project/pull/79646/commits/d5bd0215f22440e10e1a4af2b4391973831795be](https://github.com/llvm/llvm-project/pull/79646/commits/d5bd0215f22440e10e1a4af2b4391973831795be) can resolve both the issues. 

Any thoughts on this?

Reply:
> For the second argument in both cases, the current code incorrectly is showing .global .ptr even though the pointer is not pointing to objects in global memory space

Following this logic, we should not be emitting ".global" for the first argument then,as the IR does not specify the address space for it, either.

We should be consistent. If we do assume that CUDA always uses global pointers for kernel arguments, then it should apply to both arguments in the test here, regardless of  any other pointer properties.

If we do decide to be pedantic and issue address space directives only if IR contains specific AS info, then in this test case both arguments must not have `.global`. 

Same with the alignment -- alignment specification on IR level should only control `.align` directive, and nothing else.

Reply:
In the original commit linked above, the second test case does end up **not** emitting ".ptr .global .align" for the first argument since alignment is specified. For the first test case where "align 16" is specified in the first argument, we get ".ptr .global .align 16".
```
Test 1:
ptr nocapture readonly align 16 %input, ptr nocapture %out
>>> .param .u64 .ptr .global .align 16 func_param_0, .param .u64 func_param_1

Test 2:
ptr nocapture readonly %input, ptr nocapture
>>> .param .u64 func_param_0, .param .u64 func_param_1
```

We generate .ptr .global .align only when the alignment is specified.

Let me know your thoughts on this.

Reply:
> We generate .ptr .global .align only when the alignment is specified.

That's the part I do not understand. An argument being a pointer is one thing. Whether it has alignment or not is another. If it's a pointer, generate `.ptr`. If it has alignement specified, generate `.align N`.

Why would generation of `.ptr` need to depend on whether we know that pointer alignment? It does not stop being a pointer if we do not know its alignment.

Reply:
Thanks for the suggestion above. I was working on a narrowed down test case and did not handle the other cases. Will work on the pointer and alignment as mentioned and update.

Reply:
@Artem-B I have updated the attributed in this change. With the assumption that CUDA always uses global pointers for kernel arguments, .global .ptr is generated for both the arguments. The .align is generated only with the alignment is specified in CUDA, but in non CUDA alignment is default specified.

Reply:
@Artem-B can you please review this. Thanks.

Reply:
@Artem-B just a reminder for this review. Thank you.